# I. Import Library and Data

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
org_f303 = pd.read_csv('/content/drive/Shareddrives/BCTT/1. Data/f303.csv')
f303 = org_f303.copy()
f303.head()

REASON_TYPE STOPACC COVID         Số TK vay Ngày giải ngân Ngày đáo hạn  \
0       COVID       Y     x  001LHC1133570001     2013-12-23   2022-12-23   
1       COVID       Y     x  001LHC1151570001     2015-06-06   2025-06-06   
2       COVID       Y     x  001LVND153640001     2015-12-30   2025-12-30   
3       COVID       Y     x  001LVND160600002     2016-02-29   2026-02-28   
4       COVID       Y     x  001MVND160670501     2016-03-07   2026-03-07   

                            Mục đích vay cụ thể           INITIALS  \
0                           COVID1_VAY TIEU DUNG  001LHC1133570001   
1              COVID1_VAY TIEU DUNG THE CHAP BDS  001LHC1151570001   
2            COVID1_THANH TOAN CHO HD SO 0000070  001LVND153640001   
3  COVID1_TT TIEN THEO HD SO 01 TV&ĐVC 2015/HĐKT  001LVND160600002   
4           COVID1_TT TIEN TAM UNG DOT 1 CHO CTY  001MVND160670501   

   CIF_generated  
0           8528  
1           8528  
2         394361  
3         394361  
4         394361

In [ ]:
org_f304 = pd.read_csv('/content/drive/Shareddrives/BCTT/1. Data/f304.csv')
f304 = org_f304.copy()
f304.head()

Nhóm nợ thẻ tín dụng  Nhóm chốt max   Du no chot 31.12.2020  \
0                   1.0            1.0            2.569285e+08   
1                   1.0            1.0            7.137838e+05   
2                   2.0            2.0            1.827520e+07   
3                   1.0            1.0            1.262733e+08   
4                   1.0            1.0            6.791053e+06   

   Lãi quá hạn lũy kế hệ thống  Lãi quá hạn lũy kế đã thu  Lãi quá hạn lũy kế  \
0                     0.000000                   0.000000             0.00000   
1                     0.000000                   0.000000             0.00000   
2                  5642.944659                2383.031389          3259.91327   
3                     0.000000                   0.000000             0.00000   
4                          NaN                        NaN             0.00000   

   Dư nợ tính dự phòng  Tỷ lệ trích lập  \
0         2.569285e+08             0.00   
1         7.137838e+05             0.00   
2         1.827194e+07             0.05   
3                  NaN             0.00   
4         6.791053e+06             0.00   

   Dự phòng cụ thể  phải trích đến 31/12/2020  CIC  DEBT SOLD  ACCMAX  \
0                                    0.000000  0.0          0       1   
1                                    0.000000  0.0          0       1   
2                               913597.037715  0.0          0       2   
3                                    0.000000  0.0          0       1   
4                                    0.000000  NaN          0       1   

   CARDMAX  LOAN  CIFCARD  CIF_generated  
0        1     0        1              1  
1        1     0        1              2  
2        2     0        2              3  
3        1     0        1              4  
4        1     0        1              5

In [ ]:
org_f402 = pd.read_csv('/content/drive/Shareddrives/BCTT/1. Data/f402.csv')
f402 = org_f402.copy()
f402.head()

CIF  CUST_AC_NO  CCY  ACY_CURR_BALANCE  LCY_CURR_BALANCE  VALUE_DATE  \
0  385027      217553  VND      3.613481e+07      3.613481e+07  2018-06-13   
1   84335       78146  VND      1.520749e+09      1.520749e+09  2018-07-14   
2   85434      155149  VND      3.596996e+07      3.596996e+07  2018-07-25   
3   83591       89578  VND      3.925967e+08      3.925967e+08  2018-07-09   
4   84115       67428  VND      2.816612e+08      2.816612e+08  2018-07-16   

  MATURITY_DATE CUSTOMER_TYPE   RATE  STATUS  CIC  CARD  LOAN  DDMAX  
0    2022-06-13             C  22.00       4    0     0     0      4  
1    2028-07-14             I  12.20       4    0     0     0      4  
2    2022-07-25             C  20.00       2    0     0     0      2  
3    2024-07-09             I  12.70       2    0     0     0      2  
4    2024-07-16             I  12.35       2    0     0     0      2

# II. Data Preprocessing


## 2.2 F303

In [ ]:
f303.head()

REASON_TYPE STOPACC COVID         Số TK vay Ngày giải ngân Ngày đáo hạn  \
0       COVID       Y     x  001LHC1133570001     2013-12-23   2022-12-23   
1       COVID       Y     x  001LHC1151570001     2015-06-06   2025-06-06   
2       COVID       Y     x  001LVND153640001     2015-12-30   2025-12-30   
3       COVID       Y     x  001LVND160600002     2016-02-29   2026-02-28   
4       COVID       Y     x  001MVND160670501     2016-03-07   2026-03-07   

                            Mục đích vay cụ thể           INITIALS  \
0                           COVID1_VAY TIEU DUNG  001LHC1133570001   
1              COVID1_VAY TIEU DUNG THE CHAP BDS  001LHC1151570001   
2            COVID1_THANH TOAN CHO HD SO 0000070  001LVND153640001   
3  COVID1_TT TIEN THEO HD SO 01 TV&ĐVC 2015/HĐKT  001LVND160600002   
4           COVID1_TT TIEN TAM UNG DOT 1 CHO CTY  001MVND160670501   

   CIF_generated  
0           8528  
1           8528  
2         394361  
3         394361  
4         394361

### 2.2.1 Quick Overview

In [ ]:
f303_des = f303.describe()

In [ ]:
f303_null_check = f303.isnull().sum()
f303_null_check

REASON_TYPE             1036
STOPACC                 1036
COVID                      0
Số TK vay                  0
Ngày giải ngân             0
Ngày đáo hạn               1
Mục đích vay cụ thể        0
INITIALS                   0
CIF_generated              0
dtype: int64

In [ ]:
f303_duplicated_check = f303.duplicated().sum()
f303_duplicated_check

0

In [ ]:
f303_info = f303.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19287 entries, 0 to 19286
Data columns (total 9 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   REASON_TYPE           18251 non-null  object
 1   STOPACC               18251 non-null  object
 2   COVID                 19287 non-null  object
 3   Số TK vay             19287 non-null  object
 4   Ngày giải ngân        19287 non-null  object
 5   Ngày đáo hạn          19286 non-null  object
 6   Mục đích vay cụ thể   19287 non-null  object
 7   INITIALS              19287 non-null  object
 8   CIF_generated         19287 non-null  int64 
dtypes: int64(1), object(8)
memory usage: 1.3+ MB


- It can be seen that this table contains **19,287 rows** of data, with some columns **having null values**, and **no duplicate** entries.
- Some column names are in Vietnamese and **need to be renamed** accordingly.

In [ ]:
f303_row, f303_column = f303.shape
print(f"This Table contains: {f303_row} rows")
print(f"This Table contains: {f303_column} column")

This Table contains: 19287 rows
This Table contains: 9 column


### 2.2.2 Rename Column


In [ ]:
f303.rename(columns = {'Số TK vay': 'LoanAccountNumber','Ngày giải ngân': 'DisbursementDate','Ngày đáo hạn': 'MaturityDate','Mục đích vay cụ thể ': 'LoanPurposeDetail','INITIALS': 'AccountInitials' }, inplace = True)

### 2.2.3 Data Type Conversion

In [ ]:
f303['DisbursementDate'] = pd.to_datetime(f303['DisbursementDate'], errors = 'coerce')
f303['MaturityDate'] = pd.to_datetime(f303['MaturityDate'], errors = 'coerce')

In [ ]:
# Drop null
f303 = f303.dropna(subset = ['MaturityDate'])

### 2.2.4 Checking Advanced Duplicated


In [ ]:
# Check for duplicates in specific columns
cols_to_check = ['LoanAccountNumber', 'CIF_generated', 'AccountInitials']

for col in cols_to_check:
    answer = f303[col].duplicated().any()
    print(f"Does the column {col} have duplicate values? - {answer}")

Cột LoanAccountNumber có giá trị trùng lặp hay không? - False
Cột CIF_generated có giá trị trùng lặp hay không? - True
Cột AccountInitials có giá trị trùng lặp hay không? - True


- Primary key identified as 'LoanAccountNumber'

In [ ]:
# Check if the values in 'LoanAccountNumber' and 'AccountInitials' are the same
check = f303['LoanAccountNumber'] == f303['AccountInitials']
print(check.value_counts())

False    10435
True      8851
Name: count, dtype: int64

- Conclusion: The suspicion that these two columns are duplicates is not confirmed.


### 2.2.5 Preprocessing 'REASON_TYPE' and 'STOPACC'

In [ ]:
f303 = f303.fillna('NULL')